engine/src/main/battlecode/world/InternalRobot.java

```diff
@@ -392,14 +392,21 @@ private int getDamage() {
         return this.type.damage;
     }
 
+    private int locationToInt(MapLocation loc) {
+        return this.gameWorld.locationToIndex(loc);
+    }
+
     /**
-     * Attacks another robot (launcher). Assumes bot is in range.
+     * Attacks another location (launcher).
      * 
      * @param loc the location of the bot
      */
     public void attack(MapLocation loc) {
         InternalRobot bot = this.gameWorld.getRobot(loc);
-        if (!(bot == null)) {
+        if (bot == null || bot.getTeam() == this.getTeam() || bot.getType() == RobotType.HEADQUARTERS) {
+            // If robot is null, of your team, or a hq do no damage, otherwise do damage
+            this.getGameWorld().getMatchMaker().addAction(getID(), Action.LAUNCH_ATTACK, -locationToInt(loc) - 1);
```

Comment:
i'm not sure what this function does ;; but it looks like we're launching attacks on headquarters or bots of our own team?